File: sample-projects/php/src/Domain/Service/Creatures.php
<?php

declare(strict_types=1);

namespace De\Sots\CellarsAndCentaurs\Domain\Service;

use De\Sots\CellarsAndCentaurs\Domain\Model\Creature;
use De\Sots\CellarsAndCentaurs\Domain\Model\CreatureId;
use De\Sots\CellarsAndCentaurs\Domain\Model\NoSuchCreatureException;

interface Creatures
{
    public function save(Creature $creature): void;

    /**
     * @throws NoSuchCreatureException when no creature with the id lives in the dungeon
     */
    public function find(CreatureId $id): Creature;
}

File: sample-projects/rust/src/domain/service/creatures.rs
use crate::domain::model::creature::Creature;
use crate::domain::model::creature_id::CreatureId;
use crate::domain::model::no_such_creature_exception::NoSuchCreatureException;

pub trait Creatures {
    fn save(&mut self, creature: Creature);

    fn find(&self, id: &CreatureId) -> Result<Creature, NoSuchCreatureException>;
}
